### PR TITLE
20586 User Properties

### LINKED ISSUE #20586: WebSocket 2.1: Clarify the expected handling of user properties
Spec issue: `https://github.com/jakartaee/websocket/issues/235`

See "A.1. Changes Between 2.1 and 2.0" here: https://github.com/jakartaee/websocket/blob/master/spec/src/main/asciidoc/WebSocket.adoc

### PR DESCRIPTION
**Pull request by volosied**
fixes #20586

```
For server sessions, the initial contents of this Map must be a shallow copy of the user properties map returned from jakarta.websocket.server.ServerEndpointConfig#getUserProperties() at the point the jakarta.websocket.server.ServerEndpointConfig.Configurator#modifyHandshake() method exits.

For client sessions, the initial contents of this Map must be a shallow copy of the user properties map returned from EndpointConfig.getUserProperties() for the ClientEndpointConfig passed to WebSocketContainer.connectToServer(Class, ClientEndpointConfig, URI) or WebSocketContainer.connectToServer(Endpoint, ClientEndpointConfig, URI).
```

Shallow copy is accomplished by two wrappers 
 - ClientEndpointConfigCopyPerSession & ServerEndpointConfigCopyPerSession
 
Before the user properties were shared between all sessions, but, now, unique per session (due to the copy). 
 Other thing to note is that session user properties are populated by their respective EndpointConfig#getUserProperties() This allows  ServerEndpointConfig.Configurator's modifyHandshake() to modify the properties and they can be later viewed/ retrieved from the wsoc session in the endpoint. 

 
https://jakarta.ee/specifications/websocket/2.1/apidocs/client/jakarta/websocket/session#getUserProperties()

This PR also creates WebSocket internal versioning 

Need to verify CWWKH0053E is ok to use (I think is it).  CWWKH is reservered for websocket messagse, and I don't see CWWKH0053E taken anywhere. I think it's fine to use.

### CODE REVIEW

**Comment by volosied** [comment]
#build

**Comment by volosied** [comment]
#build

File: dev/com.ibm.ws.wsoc/resources/com/ibm/ws/wsoc/internal/resources/WebSockets.nlsprops [hunk 1]
```diff
@@ -222,9 +222,10 @@ endpoint.instance.error=CWWKH0051E: The server cannot access the {0} endpoint cl
...
-missingslashornull.uri=CWWKH0052E: An exception occurred during deployment/start of server Endpoint: {0} with uri: "{1}". Endpoint URI is either null or does not start with /.
+missingslashornull.uri=CWWKH0052E: An exception occurred during deployment/start of server Endpoint: {0} with uri: {1}. Endpoint URI is either null or does not start with /.
 missingslashornull.uri.explanation=ServerEndpoint is invalid because Endpoint URI is either null or does not start with / in endpoint class. WebSocket application is not available for requests. 
 missingslashornull.uri.useraction=Define non null Endpoint URI beginning with /. After you resolve the problem, deploy the application again.
 
-
-
+wsoc.feature.not.loaded.correctly=CWWKH0053E: The websocket feature did not load with an acceptable version.
+wsoc.feature.not.loaded.correctly.explanation=The websocket feature is loaded with a version that corresponds to the Java [Jakarta] EE specification of that websocket feature. The feature was loaded with no version defined.
+wsoc.feature.not.loaded.correctly.useraction=Check the server.xml file for existence of either a websocket feature. Only one should be specified.
```
Comment (jantley-ibm) [inline dev/com.ibm.ws.wsoc/resources/com/ibm/ws/wsoc/internal/resources/WebSockets.nlsprops]: "Check the server.xml file for existence of either a websocket feature."
This sentence does not make sense, I am unsure what it was supposed to say so I cannot recommend an edit.

**Comment by posmith** [comment]
L2 message review completed

**Comment by volosied** [comment]
#build

**Comment by volosied** [comment]
#build

File: dev/com.ibm.ws.wsoc/src/com/ibm/ws/wsoc/WsocUpgradeHandler.java [hunk 2]
```diff
@@ -76,6 +76,8 @@ public void init(TransportConnectionAccess access) {
         sessionImpl.setPathParameters();
         sessionImpl.signalAppOnOpen();
 
+        System.out.println(WebSocketVersionServiceManager.LOADED_SPEC_LEVEL);
+
```
Comment (volosied) [inline dev/com.ibm.ws.wsoc/src/com/ibm/ws/wsoc/WsocUpgradeHandler.java]: remove

File: dev/com.ibm.ws.wsoc/resources/com/ibm/ws/wsoc/internal/resources/WebSockets.nlsprops (hunk 1 above)
Comment (volosied) [reply dev/com.ibm.ws.wsoc/resources/com/ibm/ws/wsoc/internal/resources/WebSockets.nlsprops]: Sorry about that. Please check the new message. 
```
wsoc.feature.not.loaded.correctly.useraction=Check the server.xml file for the existence of multiple websocket features. Only one should be specified.
```

**Comment by volosied** [comment]
#build  #spawn.fullfat.buckets=io.openliberty.wsoc.2.1.internal_fat,io.openliberty.wsoc.internal_fat

File: dev/com.ibm.ws.wsoc/resources/com/ibm/ws/wsoc/internal/resources/WebSockets.nlsprops (hunk 1 above)
Comment (jantley-ibm) [reply dev/com.ibm.ws.wsoc/resources/com/ibm/ws/wsoc/internal/resources/WebSockets.nlsprops]: Looks good, thanks

**Comment by volosied** [comment]
#build #spawn.fullfat.buckets=io.openliberty.wsoc.2.1.internal_fat,io.openliberty.wsoc.internal_fat

File: dev/com.ibm.ws.wsoc/src/com/ibm/ws/wsoc/ServerEndpointConfigCopyPerSession.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    public  List<Extension> getExtensions() {
+        return this.epc.getExtensions();
+    }
+
+    public ServerEndpointConfig.Configurator getConfigurator() { 
+        return this.epc.getConfigurator();
+    }
+
+
+
```
Comment (pnicolucci) [inline dev/com.ibm.ws.wsoc/src/com/ibm/ws/wsoc/ServerEndpointConfigCopyPerSession.java]: Can you just clean up some of this white space?

File: dev/com.ibm.ws.wsoc/src/com/ibm/ws/wsoc/SessionImpl.java [hunk 4]
```diff
@@ -137,10 +137,13 @@ public void signalAppOnOpen(WsByteBuffer remainingBuf, boolean runWithDoPriv) {
 
         connLink.setEndpointManager(things.getEndpointManager());
 
-        // Not sure if we should do this right before or right after onOpen, spec is ambigioius, seems like right before is better.
-        userProperties = new HashMap<String, Object>();
-        things.setUserProperties(userProperties);
-
+        if(WebSocketVersionServiceManager.isWsoc21rHigher()){
```
Comment (pnicolucci) [inline dev/com.ibm.ws.wsoc/src/com/ibm/ws/wsoc/SessionImpl.java:140]: Please open an issue so we don't forget to fix this to be `isWsoc21orHigher`

File: dev/com.ibm.ws.wsoc/src/com/ibm/ws/wsoc/WebSocketVersionServiceManager.java [hunk 5]
```diff
@@ -108,4 +120,41 @@ protected void setWebsocketFactoryService(ServiceReference<WebSocketFactory> ref
...
+                }
+            }
+
+        } catch (Exception ex) {
+            if (tc.isDebugEnabled()) {
+                Tr.debug(tc, "Exception occured: " + ex.getCause());
+            }
+        }
+
+        // logger.logp(Level.WARNING, CLASS_NAME, "loadWsocVersion", "wsoc.feature.not.loaded.correctly");
```
Comment (pnicolucci) [inline dev/com.ibm.ws.wsoc/src/com/ibm/ws/wsoc/WebSocketVersionServiceManager.java]: Can we just remove this comment?

File: dev/com.ibm.ws.wsoc/src/com/ibm/ws/wsoc/WebSocketVersionServiceManager.java [hunk 6]
```diff
@@ -108,4 +120,41 @@ protected void setWebsocketFactoryService(ServiceReference<WebSocketFactory> ref
...
+
+        Tr.error(tc, "wsoc.feature.not.loaded.correctly");
+
+        return WebSocketVersionServiceManager.DEFAULT_VERSION;
+    }
+
+    public static boolean isWsoc21rHigher(){
```
Comment (pnicolucci) [inline dev/com.ibm.ws.wsoc/src/com/ibm/ws/wsoc/WebSocketVersionServiceManager.java:151]: Ensure we have an issue opened to fix this method spelling mistake.

File: dev/com.ibm.ws.wsoc/src/com/ibm/ws/wsoc/WsocUpgradeHandler.java [hunk 7]
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2017 IBM Corporation and others.
+ * Copyright (c) 2017, 2022 IBM Corporation and others.
  * All rights reserved. This program and the accompanying materials
```
Comment (pnicolucci) [inline dev/com.ibm.ws.wsoc/src/com/ibm/ws/wsoc/WsocUpgradeHandler.java]: Why the copyright update, did we make changes here previously and forget to update the copyright?

File: dev/io.openliberty.wsoc.2.1.internal/bnd.bnd [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+
+Bundle-Name: WebSocket
+Bundle-SymbolicName: io.openliberty.wsoc.2.1.internal
+Bundle-Description: WebSocket, version ${bVersion}
+
+Export-Package: \
+ io.openliberty.wsoc.speclevel
+
+-buildpath: \
+  com.ibm.ws.org.osgi.annotation.versioning;version=latest
```
Comment (pnicolucci) [inline dev/io.openliberty.wsoc.2.1.internal/bnd.bnd]: new line

File: dev/io.openliberty.wsoc.2.1.internal_fat/fat/src/io/openliberty/wsoc/endpoints/client/basic/ClientUserPropertiesClientEP.java [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+import jakarta.websocket.HandshakeResponse;
+import jakarta.websocket.Extension;
+import jakarta.websocket.Encoder;
+import jakarta.websocket.Decoder;
+
+import javax.net.ssl.SSLContext;
+
+/**
+ *
+ */
```
Comment (pnicolucci) [inline dev/io.openliberty.wsoc.2.1.internal_fat/fat/src/io/openliberty/wsoc/endpoints/client/basic/ClientUserPropertiesClientEP.java:58]: Did you want to add a javadoc?

File: dev/io.openliberty.wsoc.2.1.internal_fat/fat/src/io/openliberty/wsoc/tests/Basic21Test.java [hunk 10]
```diff
@@ -205,4 +207,33 @@ public void testSSCNegativeoTimeOut() throws Exception {
...
+    @Mode(TestMode.LITE)
+    @Test
+    public void testSSCUserPropertiesOnServer() throws Exception {
+        this.runAsLSAndVerifyResponse("UserPropertiesTest", "testUserPropertiesOnServer");
+    }
+
+    /* 
+     * Currently fails on Eclipse Jetty. Doesn't appear to be spec compliant.
+     * This test passes when run on Tomcat 10.1.0, however. 
+     *
```
Comment (pnicolucci) [inline dev/io.openliberty.wsoc.2.1.internal_fat/fat/src/io/openliberty/wsoc/tests/Basic21Test.java]: Do we need any issues to investigate this further or should we just remove this commented out code?

File: dev/com.ibm.ws.wsoc/src/com/ibm/ws/wsoc/SessionImpl.java (hunk 4 above)
Comment (volosied) [reply dev/com.ibm.ws.wsoc/src/com/ibm/ws/wsoc/SessionImpl.java:140]: https://github.com/OpenLiberty/open-liberty/issues/22257

File: dev/io.openliberty.wsoc.2.1.internal_fat/fat/src/io/openliberty/wsoc/tests/Basic21Test.java (hunk 10 above)
Comment (volosied) [reply dev/io.openliberty.wsoc.2.1.internal_fat/fat/src/io/openliberty/wsoc/tests/Basic21Test.java]: I can remove the comment. This problem is with Jetty, so maybe they'll fix it in a later release.

**Review by pnicolucci — APPROVED** [review]

**Comment by volosied** [comment]
#libby

**Pull request merged**